src/DependencyInjection/Compiler/ProjectorPass.php

```diff
@@ -18,67 +18,123 @@
                 );
-
-                if ($id !== sprintf('%s.%s', ProophEventStoreExtension::TAG_PROJECTION, $tag['projection_name'])) {
-                    $container->setAlias(sprintf('%s.%s', ProophEventStoreExtension::TAG_PROJECTION, $tag['projection_name']), $id);
-                }
+                $projectionsLocator[$tag['projection_name']] = new Reference($id);
             }
         }
 
-        $container
-            ->setDefinition(
-                'prooph_event_store.projection_read_models_locator',
-                new Definition(ServiceLocator::class, [$readModelsLocator])
-            )
-            ->addTag('container.service_locator');
+        self::addServicesToLocator($container, 'prooph_event_store.projections_locator', $projectionsLocator);
+        self::addServicesToLocator($container, 'prooph_event_store.projection_read_models_locator', $readModelsLocator);
+        self::addServicesToLocator(
+            $container,
+            'prooph_event_store.projection_manager_for_projections_locator',
+            $projectionManagerForProjectionsLocator
+        );
+    }
+
+    private static function addServicesToLocator(
+        ContainerBuilder $container,
+        string $locatorId,
+        array $serviceMap
+    ): void {
+        $definition = $container->getDefinition($locatorId);
+        $definition->replaceArgument(0, array_merge($serviceMap, $definition->getArgument(0)));
+    }
+
+    /**
+     * @param string $serviceId The id of the service that is verified
+     * @param ReflectionClass $projectionClass The Reflection of the service that is verified
+     * @throws RuntimeException if the service does implement neither ReadModelProjection nor Projection.
+     */
+    private static function assertProjectionHasAValidClass(string $serviceId, ReflectionClass $projectionClass): void
+    {
+        if (! $projectionClass->implementsInterface(ReadModelProjection::class)
+            && ! $projectionClass->implementsInterface(Projection::class)
+        ) {
+            throw new RuntimeException(sprintf(
+                'Tagged service "%s" must implement either "%s" or "%s"',
+                $serviceId,
+                ReadModelProjection::class,
+                Projection::class
+            ));
+        }
+    }
+
+    /**
+     * @param string $serviceId The id of the service whose tag is verified
+     * @param array $tag The actual tag
+     * @param string $attributeName The attribute that has to be available in the tag
+     * @throws RuntimeException if the attribute is not available in the tag
+     */
+    private static function assertProjectionTagHasAttribute(string $serviceId, array $tag, string $attributeName): void
+    {
+        if (! isset($tag[$attributeName])) {
+            throw new RuntimeException(sprintf(
+                '"%s" attribute is missing from on "%s" tagged service "%s"',
```

Comment:
There's a typo here.

Correct string should be stg like `"%s" attribute is missing from tag "%s" on service "%s".`